include/boost/url/detail/any_allocator.hpp

```diff
@@ -0,0 +1,251 @@
+//
+// Copyright (c) 2022 Alan Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+
+#ifndef BOOST_URL_DETAIL_ANY_ALLOCATOR_HPP
+#define BOOST_URL_DETAIL_ANY_ALLOCATOR_HPP
+
+#include <boost/url/detail/type_id.hpp>
+
+#include <boost/core/allocator_access.hpp>
+#include <boost/core/empty_value.hpp>
+#include <boost/type_traits/type_with_alignment.hpp>
+
+#include <cstddef>
+
+namespace boost {
+namespace urls {
+namespace detail {
+
+struct any_allocator_base
+{
+    struct base
+    {
+        type_id_t type_id;
+
+        virtual ~base() = default;
+
+        virtual void*
+        allocate(std::size_t n,
+            std::size_t size,
+            std::size_t align) = 0;
+
+        virtual void
+        deallocate(void* p,
+            std::size_t n,
+            std::size_t size,
+            std::size_t align) = 0;
+
+        virtual bool
+        is_equal(const any_allocator_base::base&
+                p) = 0;
+    };
+
+    template <typename Allocator>
+    struct impl
+        : public base
+        , private empty_value<
+              boost::allocator_rebind_t<Allocator,
+                  char>,
+              0>
+    {
+        using allocator_type = boost::
+            allocator_rebind_t<Allocator, char>;
+
+        explicit impl(allocator_type const& a)
+            : empty_value<allocator_type, 0>(
+                  empty_init_t(), a)
+        {
+            type_id = get_type_id<allocator_type>();
+        }
+
+        void*
+        allocate(std::size_t n,
+            std::size_t size,
+            std::size_t align) override
+        {
+            if(align == 1)
+                return allocator().allocate(n);
+            else
+                return allocate_aligned(
+                    n, size, align);
+        }
+
+        void
+        deallocate(void* p,
+            std::size_t n,
+            std::size_t size,
+            std::size_t align) override
+        {
+            if(align == alignof(char))
+                allocator().deallocate(
+                    reinterpret_cast<char*>(p),
+                    n);
+            else
+                deallocate_aligned(
+                    p, n, size, align);
+        }
+
+        bool
+        is_equal(
+            const any_allocator_base::base& other)
+            override
```

Comment:
I never break `override`, `const`, or `noexcept` to the next line, even if it would go over 50 cols